cli/command/service/logs.go

```diff
@@ -66,26 +73,91 @@ func runLogs(dockerCli *command.DockerCli, opts *logsOptions) error {
 	}
 
 	client := dockerCli.Client()
+
+	service, _, err := client.ServiceInspectWithRaw(ctx, opts.service)
+	if err != nil {
+		return err
+	}
+
 	responseBody, err := client.ServiceLogs(ctx, opts.service, options)
 	if err != nil {
 		return err
 	}
 	defer responseBody.Close()
 
-	resolver := idresolver.New(client, opts.noResolve)
+	var replicas uint64
+	if service.Spec.Mode.Replicated != nil && service.Spec.Mode.Replicated.Replicas != nil {
+		replicas = *service.Spec.Mode.Replicated.Replicas
+	}
+	padding := len(strconv.FormatUint(replicas, 10))
+
+	taskFormatter := newTaskFormatter(client, opts, padding)
 
-	stdout := &logWriter{ctx: ctx, opts: opts, r: resolver, w: dockerCli.Out()}
-	stderr := &logWriter{ctx: ctx, opts: opts, r: resolver, w: dockerCli.Err()}
+	stdout := &logWriter{ctx: ctx, opts: opts, f: taskFormatter, w: dockerCli.Out()}
+	stderr := &logWriter{ctx: ctx, opts: opts, f: taskFormatter, w: dockerCli.Err()}
 
 	// TODO(aluzzardi): Do an io.Copy for services with TTY enabled.
 	_, err = stdcopy.StdCopy(stdout, stderr, responseBody)
 	return err
 }
 
+type taskFormatter struct {
+	client  client.APIClient
+	opts    *logsOptions
+	padding int
+
+	r     *idresolver.IDResolver
+	cache map[logContext]string
+}
+
+func newTaskFormatter(client client.APIClient, opts *logsOptions, padding int) *taskFormatter {
+	return &taskFormatter{
+		client:  client,
+		opts:    opts,
+		padding: padding,
+		r:       idresolver.New(client, opts.noResolve),
+		cache:   make(map[logContext]string),
+	}
+}
+
+func (f *taskFormatter) format(ctx context.Context, logCtx *logContext) (string, error) {
```

Comment:
nit: This could just take a `logContext` argument instead of `*logContext`. It's a small structure and there's no need here for pass-by-reference semantics. Not using a pointer makes it clearer that this function will not modify `logCtx`, and the `f.cache[*logCtx] = ...` can become `f.cache[logCtx] = ...`.

I'd also suggest having `parseContext` return a `logContext` instead of `*logContext`.